x-pack/plugins/ml/public/settings/calendar/edit/import_modal.js

```diff
@@ -0,0 +1,289 @@
+        const recurring = (e.properties.RRULE !== undefined);
+
+        if (description && start && end && description.length && start.length && end.length) {
+          // Temp reference to unsaved events to allow removal from table
+          const tempId = `${i}${start[0].value.valueOf()}`;
+
+          mlEvents.push({
+            event_id: tempId,
+            description: description[0].value,
+            start_time: start[0].value.valueOf(),
+            end_time: end[0].value.valueOf(),
+            asterisk: recurring
+          });
+        }
+      }
+    });
+    return mlEvents;
+  }
+
+  static parseICSFile = (data) => {
+    const cal = icalendar.parse_calendar(data);
+    return ImportModal.createEvents(cal);
+  }
+
+  static filterEvents = (events) => {
+    const now = moment().valueOf();
+    return events.filter(e => e.start_time > now);
+  }
+
+  // move to utils?
+  static readFile = (file) => {
+    return new Promise((resolve, reject) => {
+      if (file && file.size) {
+        const reader = new FileReader();
+        reader.readAsText(file);
+
+        reader.onload = (() => {
+          return () => {
+            const data = reader.result;
+            if (data === '') {
+              reject();
+            } else {
+              resolve({ data });
+            }
+          };
+        })(file);
+      } else {
+        reject();
+      }
+    });
+  }
+
+  handleImport = async (loadedFile) => {
+    const incomingFile = loadedFile[0];
+    const errorMessage = 'Could not parse ICS file.';
+    let events = [];
+
+    if (incomingFile && incomingFile.size <= (MAX_FILE_SIZE_MB * 1000000)) {
+      this.setState({ fileLoading: true, fileLoaded: true });
+
+      try {
+        const parsedFile = await ImportModal.readFile(incomingFile);
+        events = ImportModal.parseICSFile(parsedFile.data);
+
+        this.setState({
+          allImportedEvents: events,
+          fileLoading: false,
+          errorMessage: null,
+          includePastEvents: false
+        });
+      } catch (error) {
+        console.log(errorMessage, error);
+        this.setState({ errorMessage, fileLoading: false });
+      }
+    } else if (incomingFile && incomingFile.size > (MAX_FILE_SIZE_MB * 1000000)) {
+      this.setState({ fileLoading: false, errorMessage });
+    } else {
+      this.setState({ fileLoading: false, errorMessage: null });
+    }
+  }
+
+  onEventDelete = (eventId) => {
+    this.setState(prevState => ({
+      allImportedEvents: prevState.allImportedEvents.filter(event => event.event_id !== eventId),
+    }));
+  }
+
+  onCheckboxToggle = (e) => {
+    this.setState({
+      includePastEvents: e.target.checked,
+    });
+  };
+
+  handleEventsAdd = () => {
+    const events = this.selectedEvents.map((event) => ({
+      description: event.description,
+      start_time: event.start_time,
+      end_time: event.end_time
+    }));
+
+    this.props.addImportedEvents(events);
+  }
+
+  renderCallout = () => (
```

Comment:
these `render...` functions could be react components e.g. `<ErrorCallout />` or `<ImportedEvents />`

Reply:
🤔 I think the `renderCallout` is simple enough since it simply returns the `EuiCallOut` component - I don't think we'd gain much from wrapping it in another component.

`renderImportedEvents`, on the other hand, might be complex enough to justify having a separate component. I also agree with the comment below you made - makes sense to move `<ImportedEvents />` out and avoid the setting of `selectedEvents` side-effect. 👍
Thanks!

Reply:
Thanks for catching this! Turned into a nice refactor 👌 🙌